src/dune/build_system.ml

```diff
@@ -1548,9 +1558,52 @@ end = struct
           let targets, targets_digest =
             compute_targets_digest_or_raise_error ~info targets_as_list
           in
-          Option.iter t.memory ~f:(fun memory ->
-              ignore
-                (Dune_manager.Client.promote memory targets rule_digest [] None));
+          let () =
+            (* Check cache *)
+            match from_dune_memory with
+            | Some cached when cache_checking ->
+              let rec check cached = function
+                | (file, digest) :: targets ->
+                  let cached_file, cached =
+                    List.partition
+                      ~f:(fun (cached_file : Dune_memory.File.t) ->
+                        Path.Build.equal cached_file.in_the_build_directory
+                          file)
+                      cached
+                  in
+                  ( match cached_file with
+                  | [] ->
+                    Console.print
+                      (Format.asprintf
+                         "cache mismatch on %a: missing from the cache\n"
+                         Path.Build.pp file)
+                  | [ cached_file ] ->
+                    if not (Digest.equal cached_file.digest digest) then
+                      Console.print
+                        (Format.asprintf
+                           "cache mismatch on %a: hash differ with %a\n"
+                           Path.Build.pp file Path.pp cached_file.in_the_memory)
+                  | _ ->
```

Comment:
I don't understand what this case corresponds to. We can expect that `cached` and `targets` correspond to the same list of files and so match them two by two. Do you agree?

Reply:
I have what the cache says the list of targets is, and what dune says  the list of target is. I agree it should always be the same, but just as some build may not be reproducible, I suppose someone could update a rule with a different target list and not reflect it in the hash ?

In any case, I have two lists coming from external sources, so in case there *is* a mismatch I need to either report it or `raise Code_error` or something.

Reply:
The list of targets is included in the hash, so this shouldn't happen. Reporting the error seems fine, but we can write this loop with a linear complexity rather than a quadratic one.